community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/SetPropertyPipe.scala

```diff
@@ -47,15 +53,21 @@ abstract class SetPropertyPipe[T <: PropertyContainer](src: Pipe, name: String,
   def operations(query: QueryContext): Operations[T]
   def operatorName: String
 
-  private def setProperty(context: ExecutionContext, state: QueryState, id: Long) = {
+  private def setProperty(context: ExecutionContext, state: QueryState, entityId: Long, needsExclusiveLock: Boolean) = {
     val queryContext = state.query
     val maybePropertyKey = propertyKey.id(queryContext).map(_.id) // if the key was already looked up
     val propertyId = maybePropertyKey
         .getOrElse(queryContext.getOrCreatePropertyKeyId(propertyKey.name)) // otherwise create it
     val ops = operations(queryContext)
+
+    if (needsExclusiveLock) ops.acquireExclusiveLock(entityId)
+
     val value = makeValueNeoSafe(expression(context)(state))
-    if (value == null) ops.removeProperty(id, propertyId)
-    else ops.setProperty(id, propertyId, value)
+    if (value == null) ops.removeProperty(entityId, propertyId)
+    else ops.setProperty(entityId, propertyId, value)
+
+    // TODO: Is release not needed? (Currently a write lock will be held until the transaction is committed, but will this always be true)
+    if (needsExclusiveLock) ops.acquireExclusiveLock(entityId)
```

Comment:
_This_ function acquires the lock twice. The others do acquire and release. The second acquire and the releases are unnecessary, I think.